text/0000-nested-method-calls.md

```diff
@@ -0,0 +1,824 @@
+when calling `vec.len()`. Hence the error.
+
+(In this MIR, I've included the `EndRegion` annotations that the
+current MIR borrowck relies on. In most examples, I will elide them
+unless they are needed to make a point. Also, in the future, when we
+move to NLL, those statements will not be present, and regions will be
+inferred based solely on where the references are *used*, but the
+general idea remains the same.)
+
+When you see the code desugared in that way, it should not surprise
+you that there is in fact a real danger here for code to crash if we
+just "turned off" this check (if we even could do such a thing). For
+example, consider this rather artificial Rust program:
+
+```rust
+let mut v: Vec<String> = vec![format!("Hello, ")];
+let s: String = format!("foo");
+v[0].push_str({ v.push(s); "World!" });
+//              ^^^^^^^^^ sneaky attempt to mutate `v`
+```
+
+This last line, if desugared into MIR, looks something like this;
+
+```rust
+// First evaluate `v[0]` to get a `&mut String`:
+tmp0 = &mut v;
+tmp1 = IndexMut::index_mut(tmp0, 0);
+tmp2 = tmp1;
+
+// Next, evaluate `{ v.push(s); "World!" }` block:
+tmp3 = &mut v;
+tmp4 = s;
+Vec::push(tmp3, tmp4);
+tmp5 = "World!";
+
+// Finally, invoke `push_str`:
+String::push_str(tmp2, tmp5);
+```
+
+The danger here lies in the fact that we evaluate `v[0]` into a
+reference first, but this reference could well be invalidated by the
+call to `Vec::push()` that occurs later on (which may resize the
+vector and hence change the address of its elements). The Rust type
+system naturally prevents this, however, because the first line (`tmp0
+= &mut v`) borrows `v`, and that borrow lasts until the final call to
+`push_str()`.
+
+In fact, even when the receiver is just a local variable (e.g.,
+`vec.push(vec.len())`) we have to be wary. We wouldn't want it to be
+possible to give ownership of the receiver away in one of the
+arguments: `vec.push({ send_to_another_thread(vec); ... })`. That
+should still be an error of course.
+
+(Naturally, these complex arguments that are blocks look really
+artificial, but keep in mind that most of the time when this occurs in
+practice, the argument is a method or fn call, and that could in
+principle have arbitrary side-effects.)
+
+### Introducing reservations
+
+This RFC proposes extending MIR with the concept of a **two-phase
+borrow**. These borrows are a variant of mutable borrows where the
+value starts out as **reserved** and only becomes mutably borrowed
+when the resulting reference is first used (which is called
+**activating** the borrow). During the reservation phase before a
+mutable borrow is activated, it acts exactly like a shared borrow --
+hence the borrowed value can still be read.
+
+As discussed earlier, this RFC itself only introduces these two-phase
+borrows in a limited way. Specifically, we extend the MIR with a new
+kind of borrow (written `mut2`, for two-phase), and we generate those
```

Comment:
I debated about this. I thought it was strange, because `&reserve foo` suggested that it *only* reserved `foo` (rather than later mutating it), but perhaps it's clearer to say that it produces a reserved reference that (upon first use) is "upgraded" to a regular mutable reference.